File: frontend/src/api/activityFeed.ts
import api from './axios'

export interface ActivityFeedItem {
  type: string
  timestamp: string
  actor_display_name: string | null
  actor_faction_slug: string | null
  actor_avatar_url: string | null
  payload: Record<string, any>
  /** Faction this card's frame themes to (surface #12): actor's faction, else
   *  the task's faction, else null (neutral). Derived server-side. */
  context_faction_slug: string | null
}

export interface FeedCounts {
  all: number
  friends: number
  foes: number
  your_stuff: number
  global_count: number
  requests: number
}

export interface ActivityFeedResponse {
  items: ActivityFeedItem[]
  counts: FeedCounts
  next_cursor: string | null
}

export async function getActivityFeed(params?: {
  filter?: string
  before?: string
  limit?: number
}): Promise<ActivityFeedResponse> {
  const { data } = await api.get<ActivityFeedResponse>('/activity-feed', { params })
  return data
}

File: backend/schemas/activity_feed.py
"""Schemas for the unified activity feed."""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class ActivityFeedItem(BaseModel):
    """A single item in the activity feed.

    The ``type`` field discriminates the payload shape:
    - vote_on_mine: someone voted on your submission
    - friend_completion: a friend completed a task
    - foe_taunt: a foe sent a taunt
    - global_task: a new task was activated
    - era_announcement: a new era started
    - collab_invite: someone invited you to collaborate
    - duel_challenge: someone challenged you to a duel
    - friend_signup: a friend signed up for a task you're doing
    """

    type: str
    timestamp: datetime
    actor_display_name: Optional[str] = None
    actor_faction_slug: Optional[str] = None
    actor_avatar_url: Optional[str] = None
    payload: dict[str, Any]


class FeedCounts(BaseModel):
    """Badge counts for each filter tab."""

    all: int = 0
    friends: int = 0
    foes: int = 0
    your_stuff: int = 0
    global_count: int = 0
    requests: int = 0


class ActivityFeedResponse(BaseModel):
    """Paginated activity feed with badge counts."""

    items: list[ActivityFeedItem]
    counts: FeedCounts
    next_cursor: Optional[str] = None
